src/main/java/io/jenkins/plugins/intotorecorder/transport/Grafeas.java

```diff
@@ -0,0 +1,177 @@
+
+                public Artifact(String resourceUri, Map<String, String> hashes) {
+                    this.resourceUri = resourceUri;
+                    this.hashes = hashes;
+                }
+
+            }
+
+            public List<String> command = new ArrayList<String>();
+            public List<Artifact> materials = new ArrayList<Artifact>();
+            public List<Artifact> products = new ArrayList<Artifact>();
+            public Map<String, Map<String, String>> byproducts = new HashMap<String, Map<String, String>>();
+            public Map<String, Map<String, String>> environment = new HashMap<String, Map<String, String>>();
+
+            public GrafeasInTotoLink(List<String> command,
+                                     Map<String, ArtifactHash> materials,
+                                     Map<String, ArtifactHash> products,
+                                     Map<String, Object> byproducts,
+                                     Map<String, Object> environment) {
+
+                this.command = command;
+
+                for (Map.Entry<String, ArtifactHash> material : materials.entrySet()) {
+                    // FIXME: String resourceUri = "file://sha256:" + material.getValue().get("sha256") + ":" + material.getKey();
+                    Artifact artifact = new Artifact(material.getKey(), (Map<String, String>)material.getValue());
+                    this.materials.add(artifact);
+                }
+
+                for (Map.Entry<String, ArtifactHash> product : products.entrySet()) {
+                    // FIXME: String resourceUri = "file://sha256:" + product.getValue().get("sha256") + ":" + product.getKey();
+                    Artifact artifact = new Artifact(product.getKey(), (Map<String, String>)product.getValue());
+                    this.products.add(artifact);
+                }
+
+                Map<String, String> stringByproducts = new HashMap<String, String>();
+                for (String byproduct : byproducts.keySet()) {
+                    stringByproducts.put(byproduct, String.valueOf(byproducts.get(byproduct)));
+                }
+
+                Map<String, String> stringEnvironment = new HashMap<String, String>();
+                for (String env : environment.keySet()) {
+                    stringEnvironment.put(env, String.valueOf(environment.get(env)));
+                }
+
+                this.byproducts.put("customValues", stringByproducts);
+                this.environment.put("customValues", stringEnvironment);
+            }
+        }
+
+        public GrafeasInTotoMetadata(Link link) {
+            this.signatures = link.getSignatures();
+            this.signed = new GrafeasInTotoLink(
+                link.getCommand(),
+                link.getMaterials(),
+                link.getProducts(),
+                link.getByproducts(),
+                link.getEnvironment()
+            );
+        }
+    }
+
+
+    public class GrafeasOccurrence {
```

Comment:
I'm curious, does Grafeas have a Java library? could we use that?

Reply:
https://github.com/googleapis/java-grafeas

They do have a java client, but it hasn't been updated with in-toto support yet. I wonder what the cycle is like, because in-toto is in Grafeas v1beta1, while this currently just says v1 (which also has protos defined). I agree it makes more sense to use an official library though.

Reply:
I think we could move into the library once their client is updated... For now I think things work :)